### PR TITLE
Add GitHub Actions workflow for Docker images

### PR DESCRIPTION
**Pull request by hmenke**
@the-hampel Maybe this would be desirable. It builds the Docker images on GitHub Actions and deploys them to the GitHub package registry. Layers are cached using the GitHub Actions Cache (GHA), so if the Dockerfile doesn't change, rebuilds only take a few seconds. If a rebuild is desired, the GHA can be invalidated manually.

Unfortuanately at the moment all but `github_ci` are broken (see comments in the workflow file).

### CODE REVIEW

File: .github/workflows/docker.yml
Comment (the-hampel) [inline .github/workflows/docker.yml:42]: The final docker images land on ghcr.io I assume? So we should then remove the one on docker hub?

File: .github/workflows/docker.yml
Comment (the-hampel) [inline .github/workflows/docker.yml:54]: Can we somehow tag the docker images created here with `:unstable`, so that we can create another action for 3.2.x, in which the 3.2.x branch of all apps is build (I can take care of that).

File: .github/workflows/docker.yml
Comment (the-hampel) [inline .github/workflows/docker.yml]: the jenkins_ci and triqs_default can be removed. The triqs default dockerfile can be removed completely I think and jenkins_ci is only relevant for jenkins, with docker images only available locally I think?

File: Docker/mpich_dockerfile
Comment (the-hampel) [inline Docker/mpich_dockerfile:1]: Is the mpich dockerfile relevant for you? I think otherwise we can only build the openmpi one in my opinion (+ the github_ci one)

File: Docker/github_ci_dockerfile
Comment (the-hampel) [inline Docker/github_ci_dockerfile:98]: I think github action runners have only two cores right? Maybe we should reduce `ARG NCORES=2` then. Right now it is 8 in all dockerfiles.

File: Docker/mpich_dockerfile
Comment (the-hampel) [inline Docker/mpich_dockerfile:91]: same here, unless this has changed and the github runners have now more cores.

**Review by the-hampel — COMMENTED** [review]
Thank you @hmenke for the useful addition! I made a few comments that I would like to address / discuss before merging.

File: .github/workflows/docker.yml
Comment (hmenke) [reply .github/workflows/docker.yml:42]: You can keep both. I don't see any harm.

File: .github/workflows/docker.yml
Comment (hmenke) [reply .github/workflows/docker.yml:54]: Images are automatically tagged by their branch name. You can try it out using
```
docker pull ghcr.io/hmenke/solid_dmft_github_ci:unstable
```

File: Docker/github_ci_dockerfile
Comment (hmenke) [reply Docker/github_ci_dockerfile:98]: Already covered in the workflow file:
```yaml
- name: Build and push Docker image
  uses: docker/build-push-action@v5
  with:
    build-args: |
      NCORES=2
```

File: Docker/mpich_dockerfile
Comment (hmenke) [reply Docker/mpich_dockerfile:1]: It's not relevant for me. Should I then delete `mpich_Dockerfile`?

File: .github/workflows/docker.yml
Comment (hmenke) [reply .github/workflows/docker.yml]: Should I then delete `triqs_default_Dockerfile`?

Comment (the-hampel) [reply .github/workflows/docker.yml]: yes it can go. Thanks.

File: Docker/mpich_dockerfile
Comment (the-hampel) [reply Docker/mpich_dockerfile:1]: please leave the Dockerfile itself there (your corrections are helpful), but we do not need to include it in the automatic build.

File: Docker/github_ci_dockerfile
Comment (the-hampel) [reply Docker/github_ci_dockerfile:98]: is this overwriting the `NCORES` ARG in the Dockerfile, otherwise it will still use 8 cores for the build no?

Comment (hmenke) [reply Docker/github_ci_dockerfile:98]: Yes, this is how `ARG` works. https://docs.docker.com/engine/reference/builder/#arg

Comment (the-hampel) [reply Docker/github_ci_dockerfile:98]: okay got it. Did not now the workflow passes this into the docker build. makes sense.

**Review by the-hampel — APPROVED** [review]
I think this looks good to me now. If you are also happy with the current form @hmenke I will merge this once all tests pass. Thank you for working on this!

**Comment by hmenke** [comment]
LGTM :ok_hand: Probably best to [squash and merge](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits) so that all the intermediate back and forth disappears.

**Pull request merged**
